### PR TITLE
feat: tests for IPIP-402 CAR parameters 

### PR DESCRIPTION
**Pull request by aschmahmann**
Conformance tests for IPIP-402.

- [x] Superseded and includes #49
- [ ] IPIP: https://github.com/ipfs/specs/pull/402
- [x] Boxo PR: https://github.com/ipfs/boxo/pull/303
- [x] Kubo PR: https://github.com/ipfs/kubo/pull/9914
- [x] **CI changes must be reverted before merging**: @hacdias updated this PR to run against the [Kubo PR](https://github.com/ipfs/kubo/pull/9914)

---

cc @guanzo @laurentsenta

Note: the last commits for cloning the requestbuilder and splitting out the repeatable tasks (e.g. format parameter + accept header testing) could use separate review on if they're a good idea/pattern.

### CODE REVIEW

File: tests/t0118_gateway_car_test.go [hunk 1]
```diff
@@ -13,44 +14,179 @@ func TestGatewayCar(t *testing.T) {
...
+	formatParamReq := testWithFormatParam.Request
+	expected := testWithFormatParam.Response
+
+	carFormatQueryParams, found := formatParamReq.Query_["format"]
+	if !found {
+		t.Fatal("could not find 'format' query parameter")
+	}
+
+	if len(carFormatQueryParams) != 1 {
+		t.Fatal("only using a format parameter is supported")
```
Comment (laurentsenta) [inline tests/t0118_gateway_car_test.go:160]:
```suggestion
		t.Fatal("only using a single format parameter is supported")
```

File: tests/t0118_gateway_car_test.go [hunk 2]
```diff
@@ -13,44 +14,179 @@ func TestGatewayCar(t *testing.T) {
...
+				),
+		},
 	}
 
-	Run(t, tests)
+	transformedTests := standardCARTestTransforms(t, tests)
+	Run(t, transformedTests)
+}
+
+func standardCARTestTransforms(t *testing.T, tests SugarTests) SugarTests {
```
Comment (laurentsenta) [inline tests/t0118_gateway_car_test.go:123]: That looks fine, I like the idea of using this decorator-ish approach,

We do call stuff like `Clone()` and `delete(map, k)` which moves away from our builder pattern and digs quite deep into the internals: could you move this into `tooling/helpers`?

File: tests/t0118_gateway_car_test.go [hunk 3]
```diff
@@ -13,44 +14,179 @@ func TestGatewayCar(t *testing.T) {
...
+		},
+		{
+			Name: "GET CAR with dag-scope=entity",
+			Hint: `
+				dag-scope=entity should return a CAR file with all the blocks needed to 'cat'
+				a UnixFS file at the end of the specified path, or to 'ls' a UnixFS directory
+				at the end of the specified path.
+			`,
+			Request: Request().
+				Path("ipfs/%s", fixture.MustGetCid()).
```
Comment (laurentsenta) [inline tests/t0118_gateway_car_test.go:78]:
```suggestion
				Path("ipfs/{{cid}}", fixture.MustGetCid()).
```

File: tests/t0118_gateway_car_test.go [hunk 4]
```diff
@@ -13,44 +14,179 @@ func TestGatewayCar(t *testing.T) {
...
+				),
+		},
+		{
+			Name: "GET CAR with dag-scope=all",
+			Hint: `
+				dag-scope=all should return a CAR file with the entire contiguous DAG
+				that begins at the end of the path query, after blocks required to verify path segments.
+			`,
+			Request: Request().
+				Path("ipfs/%s/subdir", fixture.MustGetCid()).
```
Comment (laurentsenta) [inline tests/t0118_gateway_car_test.go:100]:
```suggestion
				Path("ipfs/{{cid}}/subdir", fixture.MustGetCid()).
```

**Review by laurentsenta — CHANGES_REQUESTED** [review]
really cool, thanks for the PR,

A note on moving file, and fixes, the `Master` test should pass, it's [broken because of templating right now.](https://github.com/ipfs/gateway-conformance/actions/runs/5047717583?pr=56)

**Comment by laurentsenta** [comment]
Thanks for the update, I can see it's failing because of a few errors: https://github.com/ipfs/gateway-conformance/actions/runs/5060325462?pr=56#summary-13700442847

> Error: Body length mismatch: 2 != 3
> Error: Header 'Content-Disposition' expect to find substring 'attachment; filename="bafybeiefu3d7oytdumk5v7gn6s7whpornueaw7m7u46v2o6omsqcrhhkzi...
> Error: Header 'Content-Length' expected empty string, got '221' (CAR is streamed, gateway may not have the entire thing, unable to calculate total size)

The issue with Content-Length is a known bug with Kubo https://github.com/ipfs/kubo/issues/9651 we have a workaround, add the tests in https://github.com/ipfs/gateway-conformance/blob/c32f74d4d729b89b732e0943cf817abc1e1e3145/.github/workflows/test.yml#L65

Not sure about the two others, is this a new bug?

**Comment by hacdias** [comment]
Notes:

- [ ] We have a lot of `.InThatOrder` - we may want to revisit that to check.

**Comment by hacdias** [comment]
@lidel @aschmahmann tests are passing against `ipip-402` branch in Kubo, which uses the latest stuff in Boxo. I think we have a very good test coverage here.

File: tests/t0118_gateway_car_test.go [hunk 5]
```diff
@@ -5,52 +5,596 @@ import (
...
+							subdirTwoSingleBlockFilesFixture.MustGetCid(),
+							subdirTwoSingleBlockFilesFixture.MustGetCid("subdir"),
+							subdirTwoSingleBlockFilesFixture.MustGetCid("subdir", "ascii.txt"),
+						).
+						Exactly().
+						InThatOrder(),
+				),
+		},
+		{
+			Name: "GET default CAR response of UnixFS file on a path with sharded directory",
```
Comment (lidel) [inline tests/t0118_gateway_car_test.go:43]:
```suggestion
			Name: "GET default CAR response of UnixFS file on a path with HAMT-sharded directory",
```

File: tooling/car/unixfs.go [hunk 6]
```diff
@@ -164,6 +186,62 @@ func (d *UnixfsDag) MustGetChildrenCids(names ...string) []string {
...
+func (d *UnixfsDag) MustGetCIDsInHAMTTraversal(path []string, child string) []string {
+	node := d.MustGetNode(path...)
+	var cids []string
+	tracker := dservTrackingWrapper{
+		DAGService: node.dsvc,
+	}
+	h, err := hamt.NewHamtFromDag(&tracker, node.node)
+	if err != nil {
+		panic(err)
+	}
```
Comment (lidel) [inline tooling/car/unixfs.go:234]: I am concerned how future-proof this is, especially when we have no UnixFS specs yet.
We delegate "what HAMT is" to external library which may change at any time, changing the scope of accepted behaviors.

Let's [copy these explicit asserts](https://github.com/ipfs/boxo/blob/e2fc7f2fd0237afad200d7b0eec8b7a60bdc6644/ipld/unixfs/hamt/hamt.go#LL133C1-L149C3) here, so they are hardcoded here and executed before hamt library:

```suggestion
   pbnd, ok := node.node.(*dag.ProtoNode)
	if !ok {
		return nil, dag.ErrNotProtobuf
	}

	fsn, err := format.FSNodeFromBytes(pbnd.Data())
	if err != nil {
		return nil, err
	}

	if fsn.Type() != format.THAMTShard {
		return nil, fmt.Errorf("node was not a dir shard")
	}

	if fsn.HashType() != HashMurmur3 {
		return nil, fmt.Errorf("only murmur3 supported as hash function")
	}

	h, err := hamt.NewHamtFromDag(&tracker, node.node)
	if err != nil {
		panic(err)
	}
```

File: tests/t0118_gateway_car_test.go [hunk 7]
```diff
@@ -5,52 +5,596 @@ import (
...
-					Exactly().
-					InThatOrder(),
-			),
+				Path("/ipfs/{{cid}}/subdir/ascii.txt", subdirTwoSingleBlockFilesFixture.MustGetCid()).
+				Query("format", "car"),
+			Response: Expect().
+				Status(200).
+				Body(
+					IsCar().
+						HasRoot(subdirTwoSingleBlockFilesFixture.MustGetCid()).
```
[the remark this answers is not in the document]
Comment (hacdias) [reply tests/t0118_gateway_car_test.go:32]: I don't have very strong opinions, but considering the current behaviour of `ipfs dag import`, I think we can easily change the root of the CAR to be the last element in the path. That would allow for a better UX.

File: tests/t0118_gateway_car_test.go [hunk 8]
```diff
@@ -5,52 +5,596 @@ import (
...
+			Name: "GET default CAR response with pathing through UnixFS Directory",
 			Hint: `
 				CAR stream is not deterministic, as blocks can arrive in random order,
 				but if we have a small file that fits into a single block, and export its CID
 				we will get a CAR that is a deterministic array of bytes.
 			`,
 			Request: Request().
-				Path("/ipfs/{{cid}}/subdir/ascii.txt", fixture.MustGetCid()).
-				Headers(
-					Header("Accept", "application/vnd.ipld.car"),
```
[the remark this answers is not in the document]
Comment (hacdias) [reply tests/t0118_gateway_car_test.go:25]: All tests are run with `?format` and `Accept`. This is done via `StandardCARTestTransforms` which transforms the tests.

File: tooling/car/unixfs.go (hunk 6 above)
Comment (hacdias) [reply tooling/car/unixfs.go:234]: @lidel the problem with this is that we then have to copy more than the verification: we have to copy many parts from the `Shard` struct defined in the library, in order to perform the remaining operations. Are you sure? That'll likely be quite a bit of code.

File: tests/t0118_gateway_car_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (lidel) [reply tests/t0118_gateway_car_test.go:32]: I've added 4 options to IPIP-402, let's continue there: https://github.com/ipfs/specs/pull/402#pullrequestreview-1465206498

File: tooling/car/unixfs.go (hunk 6 above)
Comment (hacdias) [reply tooling/car/unixfs.go:234]: I'll check

Comment (hacdias) [reply tooling/car/unixfs.go:255]: [`9f9652f` (#56)](https://github.com/ipfs/gateway-conformance/pull/56/commits/9f9652fa334bf5f404861b4ca0208f3567712368)

File: tests/t0118_gateway_car_test.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (lidel) [reply tests/t0118_gateway_car_test.go:32]: @hacdias @aschmahmann  based on feeback from Rod, I've relaxed requirement: https://github.com/ipfs/specs/pull/402#discussion_r1221517832
Means we need to update  conformance tests and expect either terminus CID (current behavior), or empty roots.

**Review by lidel — CHANGES_REQUESTED** [review]

File: tests/t0118_gateway_car_test.go [hunk 9]
```diff
@@ -5,52 +5,596 @@ import (
...
+				Body(
+					IsCar().
+						HasRoot(subdirTwoSingleBlockFilesFixture.MustGetCid()).
+						HasBlocks(
+							subdirTwoSingleBlockFilesFixture.MustGetCid(),
+							subdirTwoSingleBlockFilesFixture.MustGetCid("subdir"),
+							subdirTwoSingleBlockFilesFixture.MustGetCid("subdir", "ascii.txt"),
+						).
+						Exactly().
+						InThatOrder(),
```
[the remark this answers is not in the document]
Comment (hacdias) [reply tests/t0118_gateway_car_test.go:39]: @lidel it's easy to just remove it if needed.

File: tests/t0118_gateway_car_test.go [hunk 10]
```diff
@@ -5,52 +5,595 @@ import (
...
+		{
+			Name: "GET CAR with entity-bytes equivalent to a HTTP Range Request for the middle of a file",
+			Hint: `
+				The response MUST contain only the minimal set of blocks necessary for fulfilling the range request
+			`,
+			Request: Request().
+				Path("/ipfs/{{cid}}", subdirWithMixedBlockFiles.MustGetCidWithCodec(0x70, "subdir", "multiblock.txt")).
+				Query("format", "car").
+				Query("dag-scope", "entity").
+				Query("entity-bytes", "512:1023"),
```
Comment (lidel) [inline tests/t0118_gateway_car_test.go:469]: :thought_balloon: this works for all-in-one implementations like boxo/gateway, but won't be enough for Saturn.  I don't want to block on this – filled https://github.com/ipfs/gateway-conformance/issues/75 as a follow-up work.

**Review by lidel — CHANGES_REQUESTED** [review]
@aschmahmann  @hacdias fyi  I've filled https://github.com/ipfs/gateway-conformance/issues/75 to unblock landing this sooner than later.

iiuc the remaining asks are:

- [ ] update boxo code and assets for value in `CarV1Header.roots`, as noted in https://github.com/ipfs/specs/pull/402#discussion_r1221517832
- [ ] add regression test for `MustGetCIDsInHAMT*` helpers  (or fill issue for future PR)

I'm out Thu-Fri, but if above two are handled, lgtm to merge and ship.

**Comment by hacdias** [comment]
This is blocked. Tests work well if the gateway we're testing against sends roots. If not, `car.NewReader` will scream at you saying there's no roots: https://github.com/ipfs/boxo/blob/aa12ba3e27965674d3e7cef7fefdccfa8d978d9c/ipld/car/car.go#L137-L139 (same in `ipld/go-car`, which I would prefer to use here).

The CAR specification also mentions the CAR requires 1 or more roots: https://ipld.io/specs/transport/car/carv1/#constraints

We have to reach a consensus on what to do about the CAR roots, then **update the libraries if we decide 0 roots is possible**. Release those libraries and then update here.

File: tooling/car/unixfs.go [hunk 11]
```diff
@@ -164,6 +187,82 @@ func (d *UnixfsDag) MustGetChildrenCids(names ...string) []string {
 	return cids
 }
 
+// MustGetCidsInHAMT returns the cids in the HAMT at the given path. Does not include the CID of the HAMT root
+func (d *UnixfsDag) MustGetCidsInHAMT(names ...string) []string {
```
[the remark this answers is not in the document]
Comment (hacdias) [reply tooling/car/unixfs.go:190]: Added. I used the same fixture as we have for the gateway. Note that I had to do some manual inspection for the known CIDs.

**Comment by hacdias** [comment]
From standup: we decided to merge implementation in Boxo and Kubo, then this.

The decision on the CAR roots is deferred and both Boxo and conformance tests can be later updated.

⚠️ This tests currently check for 1 root (last segment) or no roots. The test logic is correct according tot he current IPIP-402 state. However, `go-car` does not support zero roots and therefore the tests will fail. Let's continue https://github.com/ipfs/specs/pull/402#discussion_r1222925940.
